File: packages/vocra/ios/Classes/VoiceFlutterPlugin.swift
import Flutter
import UIKit

/// Registers the optional native-AEC mic channels (spec §9, T18). v1's
/// happy path (`FlutterMicSource` via the `record` plugin) does not depend
/// on this class at all — it's only loaded when an app opts into
/// full-duplex mode and constructs a `NativeAecMicSource`.
public class VoiceFlutterPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    private let engine = AecAudioEngine()
    private var eventSink: FlutterEventSink?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = VoiceFlutterPlugin()

        let methodChannel = FlutterMethodChannel(
            name: "voice_flutter/aec_mic",
            binaryMessenger: registrar.messenger()
        )
        registrar.addMethodCallDelegate(instance, channel: methodChannel)

        let eventChannel = FlutterEventChannel(
            name: "voice_flutter/aec_mic/stream",
            binaryMessenger: registrar.messenger()
        )
        eventChannel.setStreamHandler(instance)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "isAvailable":
            result(AecAudioEngine.isAvailable())
        case "start":
            guard eventSink != nil else {
                result(
                    FlutterError(
                        code: "NO_LISTENER",
                        message: "Subscribe to the event stream before start()",
                        details: nil
                    )
                )
                return
            }
            do {
                try engine.start { [weak self] data in
                    // Hop to the main thread: this closure is invoked from
                    // AVAudioEngine's real-time audio thread, but
                    // FlutterEventSink must only be called on the main
                    // thread.
                    DispatchQueue.main.async {
                        self?.eventSink?(FlutterStandardTypedData(bytes: data))
                    }
                }
                result(nil)
            } catch {
                result(FlutterError(code: "START_FAILED", message: error.localizedDescription, details: nil))
            }
        case "stop":
            engine.stop()
            result(nil)
        case "sampleRate":
            result(Int(AecAudioEngine.sampleRate))
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        engine.stop()
        return nil
    }
}

File: packages/vocra_flutter/android/src/main/kotlin/com/vocra/voice_flutter/VoiceFlutterPlugin.kt
package com.vocra.voice_flutter

import android.os.Handler
import android.os.Looper
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

private const val METHOD_CHANNEL = "voice_flutter/aec_mic"
private const val EVENT_CHANNEL = "voice_flutter/aec_mic/stream"

/**
 * Registers the optional native-AEC mic channels (spec §9, T18). v1's happy path
 * ([FlutterMicSource] via the `record` plugin) does not depend on this class at all — it's only
 * loaded when an app opts into [DuplexMode.fullDuplex] and constructs a `NativeAecMicSource`.
 */
class VoiceFlutterPlugin :
    FlutterPlugin,
    MethodChannel.MethodCallHandler,
    EventChannel.StreamHandler {
    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private val mainHandler = Handler(Looper.getMainLooper())
    private val recorder = AecAudioRecorder()
    private var eventSink: EventChannel.EventSink? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel = MethodChannel(binding.binaryMessenger, METHOD_CHANNEL)
        methodChannel.setMethodCallHandler(this)
        eventChannel = EventChannel(binding.binaryMessenger, EVENT_CHANNEL)
        eventChannel.setStreamHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        recorder.stop()
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "isAvailable" -> result.success(AecAudioRecorder.isAvailable())
            "start" -> {
                val sink = eventSink
                if (sink == null) {
                    result.error("NO_LISTENER", "Subscribe to the event stream before start()", null)
                    return
                }
                try {
                    recorder.start(sink) { action -> mainHandler.post(action) }
                    result.success(null)
                } catch (e: Exception) {
                    result.error("START_FAILED", e.message, null)
                }
            }
            "stop" -> {
                recorder.stop()
                result.success(null)
            }
            "sampleRate" -> result.success(AecAudioRecorder.SAMPLE_RATE)
            else -> result.notImplemented()
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
        recorder.stop()
    }
}
